File: android/app/src/main/java/com/clipboardsync/android/filetransfer/FileTransferCrypto.kt
package com.clipboardsync.android.filetransfer

import android.util.Base64
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.security.SecureRandom
import kotlin.math.ceil

/**
 * Chunked AES-256-GCM encryption / decryption for file transfers.
 *
 * Wire format (identical to macOS FileTransferCrypto):
 *   [4 bytes magic "FTF!" = 0x46544621]
 *   [4 bytes chunk_count, big-endian]
 *   For each chunk:
 *     [4 bytes chunk_data_len, big-endian]   (= 12 nonce + ciphertext + 16 tag)
 *     [12 bytes nonce]
 *     [chunk_data_len - 12 bytes: ciphertext + GCM tag]
 */
object FileTransferCrypto {

    private const val ALGORITHM = "AES/GCM/NoPadding"
    private const val NONCE_LEN = 12       // 96-bit GCM nonce
    private const val TAG_LEN   = 128      // GCM tag length in bits
    const val CHUNK_SIZE        = 1 * 1024 * 1024  // 1 MB
    private const val MAGIC     = 0x46544621.toInt()

    private val rng = SecureRandom()

    // ── Per-transfer key helpers ─────────────────────────────────────────────

    /** Generate a fresh random 256-bit AES key for one file transfer. */
    fun generateFileKey(): ByteArray = ByteArray(32).also { rng.nextBytes(it) }

    /**
     * Wrap [fileKeyBytes] inside an AES-GCM envelope keyed by [roomKeyBytes].
     * Returns (ciphertextBase64, ivBase64) for embedding in FILE_OFFER.
     */
    fun encryptFileKey(fileKeyBytes: ByteArray, roomKeyBytes: ByteArray): Pair<String, String> {
        val nonce = ByteArray(NONCE_LEN).also { rng.nextBytes(it) }
        val cipher = aesCipher(Cipher.ENCRYPT_MODE, roomKeyBytes, nonce)
        val ct = cipher.doFinal(fileKeyBytes)
        return Base64.encodeToString(ct, Base64.NO_WRAP) to
               Base64.encodeToString(nonce, Base64.NO_WRAP)
    }

    /**
     * Unwrap and return the raw file key bytes from FILE_OFFER signal fields.
     */
    fun decryptFileKey(ciphertextBase64: String, ivBase64: String, roomKeyBytes: ByteArray): ByteArray {
        val ct    = Base64.decode(ciphertextBase64, Base64.NO_WRAP)
        val nonce = Base64.decode(ivBase64, Base64.NO_WRAP)
        val cipher = aesCipher(Cipher.DECRYPT_MODE, roomKeyBytes, nonce)
        return cipher.doFinal(ct)
    }

    // ── File Encryption → temp file (for R2 upload) ──────────────────────────

    /**
     * Encrypt [source] chunk-by-chunk into a new temp file.
     * Returns the temp File (caller must delete it).
     * [progress] callback receives 0.0..1.0.
     */
    fun encryptToTempFile(
        source: File,
        keyBytes: ByteArray,
        progress: ((Double) -> Unit)? = null
    ): File {
        val totalChunks = calcChunks(source.length())
        val temp = File.createTempFile("ftf_enc_", ".ftf")

        source.inputStream().buffered().use { inp ->
            FileOutputStream(temp).buffered().use { out ->
                out.writeInt32BE(MAGIC)
                out.writeInt32BE(totalChunks)

                var done = 0
                val buf = ByteArray(CHUNK_SIZE)
                var n: Int
                while (inp.read(buf).also { n = it } > 0) {
                    val plain = buf.copyOf(n)
                    val nonce = ByteArray(NONCE_LEN).also { rng.nextBytes(it) }
                    val cipher = aesCipher(Cipher.ENCRYPT_MODE, keyBytes, nonce)
                    val ct = cipher.doFinal(plain)               // ciphertext + 16-byte tag

                    out.writeInt32BE(NONCE_LEN + ct.size)
                    out.write(nonce)
                    out.write(ct)

                    done++
                    progress?.invoke(done.toDouble() / totalChunks)
                }
            }
        }
        return temp
    }

    // ── File Decryption ← temp file (from R2 / LAN) ─────────────────────────

    /**
     * Decrypt the FTF-format [source] file into [destination].
     * [progress] callback receives 0.0..1.0.
     */
    fun decryptFromFile(
        source: File,
        keyBytes: ByteArray,
        destination: File,
        progress: ((Double) -> Unit)? = null
    ) {
        FileInputStream(source).buffered().use { inp ->
            val magic = inp.readInt32BE()
            require(magic == MAGIC) { "Invalid FTF magic (got 0x${magic.toString(16)})" }
            val totalChunks = inp.readInt32BE()

            FileOutputStream(destination).buffered().use { out ->
                for (i in 0 until totalChunks) {
                    val chunkLen = inp.readInt32BE()
                    require(chunkLen > NONCE_LEN + 16) { "Chunk too small at index $i" }

                    val nonce = inp.readNBytes(NONCE_LEN)
                    val ct    = inp.readNBytes(chunkLen - NONCE_LEN)

                    val cipher = aesCipher(Cipher.DECRYPT_MODE, keyBytes, nonce)
                    out.write(cipher.doFinal(ct))

                    progress?.invoke((i + 1).toDouble() / totalChunks)
                }
            }
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    fun toSecretKey(keyBytes: ByteArray): SecretKey = SecretKeySpec(keyBytes, "AES")

    private fun aesCipher(mode: Int, keyBytes: ByteArray, nonce: ByteArray): Cipher {
        val key  = SecretKeySpec(keyBytes, "AES")
        val spec = GCMParameterSpec(TAG_LEN, nonce)
        return Cipher.getInstance(ALGORITHM).apply { init(mode, key, spec) }
    }

    private fun calcChunks(fileSize: Long) =
        maxOf(1, ceil(fileSize.toDouble() / CHUNK_SIZE).toInt())

    private fun java.io.OutputStream.writeInt32BE(v: Int) {
        write((v ushr 24) and 0xFF)
        write((v ushr 16) and 0xFF)
        write((v ushr  8) and 0xFF)
        write( v          and 0xFF)
    }

    private fun java.io.InputStream.readInt32BE(): Int {
        val b0 = read(); val b1 = read(); val b2 = read(); val b3 = read()
        if (b3 < 0) throw java.io.EOFException()
        return (b0 shl 24) or (b1 shl 16) or (b2 shl 8) or b3
    }

    private fun java.io.InputStream.readNBytes(n: Int): ByteArray {
        val buf = ByteArray(n)
        var off = 0
        while (off < n) {
            val r = read(buf, off, n - off)
            if (r < 0) throw java.io.EOFException("Unexpected EOF at offset $off (need $n bytes)")
            off += r
        }
        return buf
    }
}

File: macos/Sources/FileTransfer/FileTransferCrypto.swift
import Foundation
import CryptoKit

// MARK: - Error

public enum FileTransferError: LocalizedError {
    case keyNotAvailable
    case invalidPayload
    case encryptionFailed
    case decryptionFailed(String)
    case ioError(String)
    case uploadFailed(Int)
    case downloadFailed(Int)
    case transferExpired
    case invalidToken

    public var errorDescription: String? {
        switch self {
        case .keyNotAvailable:      return "Encryption key not available"
        case .invalidPayload:       return "Invalid or corrupted transfer payload"
        case .encryptionFailed:     return "File encryption failed"
        case .decryptionFailed(let m): return "File decryption failed: \(m)"
        case .ioError(let m):       return "I/O error: \(m)"
        case .uploadFailed(let c):  return "Upload failed (HTTP \(c))"
        case .downloadFailed(let c): return "Download failed (HTTP \(c))"
        case .transferExpired:      return "Transfer link has expired"
        case .invalidToken:         return "Invalid transfer token"
        }
    }
}

// MARK: - FileTransferCrypto

public final class FileTransferCrypto {

    // 1 MB chunks — memory-efficient for files up to 500 MB
    public static let chunkSize = 1 * 1024 * 1024

    // Wire format magic "FTF!" — lets receiver verify it has the right payload
    static let magic: UInt32 = 0x46544621

    // MARK: Per-transfer Key Helpers

    /// Generate a fresh random AES-256 key scoped to one file transfer.
    public static func generateFileKey() -> SymmetricKey {
        SymmetricKey(size: .bits256)
    }

    /// Wrap the per-transfer `fileKey` inside an AES-GCM envelope keyed by the room `roomKey`.
    /// Returns (ciphertextBase64, ivBase64) for embedding in the FILE_OFFER signal.
    public static func encryptFileKey(
        _ fileKey: SymmetricKey,
        with roomKey: SymmetricKey
    ) throws -> (ciphertextBase64: String, ivBase64: String) {
        let keyData = fileKey.withUnsafeBytes { Data($0) }
        let box = try AES.GCM.seal(keyData, using: roomKey)
        let iv  = box.nonce.withUnsafeBytes { Data($0).base64EncodedString() }
        let ct  = (box.ciphertext + box.tag).base64EncodedString()
        return (ct, iv)
    }

    /// Unwrap and return a `SymmetricKey` from the FILE_OFFER signal fields.
    public static func decryptFileKey(
        ciphertextBase64: String,
        ivBase64: String,
        with roomKey: SymmetricKey
    ) throws -> SymmetricKey {
        guard let ct = Data(base64Encoded: ciphertextBase64),
              let iv = Data(base64Encoded: ivBase64),
              ct.count > 16 else { throw FileTransferError.invalidPayload }

        let nonce  = try AES.GCM.Nonce(data: iv)
        let tagIdx = ct.count - 16
        let box    = try AES.GCM.SealedBox(nonce: nonce,
                                            ciphertext: ct[..<tagIdx],
                                            tag: ct[tagIdx...])
        let keyData = try AES.GCM.open(box, using: roomKey)
        return SymmetricKey(data: keyData)
    }

    // MARK: File Encryption → Temp File (for R2 upload)

    /// Encrypt `sourceURL` chunk-by-chunk, writing to a temporary `.ftf` file.
    /// Returns the URL of the temp file (caller must delete it when done).
    ///
    /// Wire format:
    /// ```
    /// [4 bytes magic][4 bytes chunk_count]
    /// [4 bytes chunk_data_len][12 bytes nonce][chunk_data_len-12 bytes ciphertext+tag] × N
    /// ```
    public static func encryptToTempFile(
        sourceURL: URL,
        key: SymmetricKey,
        progress: ((Double) -> Void)? = nil
    ) throws -> URL {
        let inHandle  = try FileHandle(forReadingFrom: sourceURL)
        defer { try? inHandle.close() }

        let fileSize    = fileBytes(at: sourceURL)
        let totalChunks = max(1, Int(ceil(Double(fileSize) / Double(chunkSize))))
        let tempURL     = FileManager.default.temporaryDirectory
                              .appendingPathComponent(UUID().uuidString + ".ftf")

        FileManager.default.createFile(atPath: tempURL.path, contents: nil)
        let outHandle = try FileHandle(forWritingTo: tempURL)
        defer { try? outHandle.close() }

        // Header
        outHandle.write(uint32BE(magic))
        outHandle.write(uint32BE(UInt32(totalChunks)))

        var chunksDone = 0
        while true {
            let plain = readChunk(from: inHandle, maxBytes: chunkSize)
            guard !plain.isEmpty else { break }

            let box    = try AES.GCM.seal(plain, using: key)
            let nonce  = box.nonce.withUnsafeBytes { Data($0) }
            let cipher = box.ciphertext + box.tag

            outHandle.write(uint32BE(UInt32(nonce.count + cipher.count)))
            outHandle.write(nonce)
            outHandle.write(cipher)

            chunksDone += 1
            progress?(Double(chunksDone) / Double(totalChunks))
        }

        return tempURL
    }

    // MARK: File Decryption ← Temp File (from R2 download)

    /// Decrypt an `.ftf` file at `encryptedURL` to `destinationURL`.
    public static func decryptFromFile(
        encryptedURL: URL,
        key: SymmetricKey,
        destinationURL: URL,
        progress: ((Double) -> Void)? = nil
    ) throws {
        let inHandle = try FileHandle(forReadingFrom: encryptedURL)
        defer { try? inHandle.close() }

        // Validate header
        let header = inHandle.readData(ofLength: 8)
        guard header.count == 8 else { throw FileTransferError.invalidPayload }

        let gotMagic = uint32FromBE(header[0..<4])
        guard gotMagic == magic else { throw FileTransferError.invalidPayload }
        let totalChunks = Int(uint32FromBE(header[4..<8]))

        FileManager.default.createFile(atPath: destinationURL.path, contents: nil)
        let outHandle = try FileHandle(forWritingTo: destinationURL)
        defer { try? outHandle.close() }

        for i in 0..<totalChunks {
            let sizeBuf = inHandle.readData(ofLength: 4)
            guard sizeBuf.count == 4 else { throw FileTransferError.invalidPayload }
            let chunkLen = Int(uint32FromBE(sizeBuf[0..<4]))

            guard chunkLen > 28 else { throw FileTransferError.invalidPayload }
            let chunkData = inHandle.readData(ofLength: chunkLen)
            guard chunkData.count == chunkLen else { throw FileTransferError.invalidPayload }

            let nonce  = try AES.GCM.Nonce(data: chunkData[0..<12])
            let tagIdx = chunkData.count - 16
            let box    = try AES.GCM.SealedBox(nonce: nonce,
                                                ciphertext: chunkData[12..<tagIdx],
                                                tag: chunkData[tagIdx...])
            let plain  = try AES.GCM.open(box, using: key)
            outHandle.write(plain)

            progress?(Double(i + 1) / Double(totalChunks))
        }
    }

    // MARK: LAN Streaming Helpers

    /// Read one chunk from a FileHandle, or return empty Data at EOF.
    public static func readChunk(from handle: FileHandle, maxBytes: Int) -> Data {
        if #available(macOS 10.15.4, *) {
            return (try? handle.read(upToCount: maxBytes)) ?? Data()
        } else {
            return handle.readData(ofLength: maxBytes)
        }
    }

    // MARK: Utility

    static func fileBytes(at url: URL) -> Int {
        (try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int) ?? 0
    }

    static func uint32BE(_ v: UInt32) -> Data {
        var be = v.bigEndian
        return Data(bytes: &be, count: 4)
    }

    static func uint32FromBE(_ slice: Data) -> UInt32 {
        slice.withUnsafeBytes { UInt32(bigEndian: $0.load(as: UInt32.self)) }
    }
}
